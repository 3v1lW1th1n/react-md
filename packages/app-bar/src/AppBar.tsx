import * as React from "react";
import * as PropTypes from "prop-types";
import cn from "classnames";

export type AppBarPosition = "top" | "bottom";
export type AppBarTheme = "clear" | "primary" | "secondary" | "default";

/**
 * The `AppBar` component is normally used to create a fixed header element within your page that defines a title,
 * optional actions, and an optional navigation button.
 */
export interface IAppBarProps extends React.HTMLAttributes<HTMLDivElement> {
  /**
   * Boolean if the `AppBar` should be fixed to the top or bottom of the page.
   *
   * @docgen
   */
  fixed?: boolean;

  /**
   * The position within the page to "fix" the `AppBar` when the `fixed` prop is enabled.
   *
   * @docgen
   */
  fixedPosition?: AppBarPosition;

  /**
   * Boolean if the fixed `AppBar` should gain elevation. This is recommended to stay enabled unless
   * you manually apply a border to help separate the `AppBar` from other content.
   *
   * @docgen
   */
  fixedElevation?: boolean;

  /**
   * Boolean if the `AppBar` should use the `dense` spec. This prop can be used along with the `prominent` prop
   * to create a prominent and dense `AppBar`.
   *
   * @docgen
   */
  dense?: boolean;

  /**
   * Boolean if the `AppBar` should use the `prominent` spec. This prop can be used along with the `dense` prop
   * to create a prominent and dense `AppBar`.
   *
   * @docgen
   */
  prominent?: boolean;

  /**
   * The theme to apply to the `AppBar`.
   *
   * @docgen
   */
  theme?: AppBarTheme;

  /**
   * Boolean if the `AppBarNav`, `AppBarTitle`, and `AppBarActions` should inherit the color that for the provided
   * `theme`. If this value is `undefined`, the color will only be inherited when the theme is set to `primary` or
   * `secondary`. However if this value is a boolean, it will be used instead. So if you set this to `false` and set
   * the `theme` to `"primary"`, the defined primary text clor will not be inherited.
   *
   * @docgen
   */
  inheritColor?: boolean;
}

export interface IAppBarDefaultProps {
  fixed: boolean;
  fixedPosition: AppBarPosition;
  fixedElevation: boolean;
  dense: boolean;
  prominent: boolean;
  theme: AppBarTheme;
}

export type AppBarWithDefaultProps = IAppBarProps & IAppBarDefaultProps;

/**
 * The `AppBar` component is usually used to create a fixed header within your page that has a title, an optional nav,
 * and optional actions. Since it is fixed on the page, it normally requires adding padding or margin to relative
 * elements so that they aren't covered by this component. You can use the static class names on the
 * `AppBar` to correctly add the padding or margin.
 *
 * ```tsx
 * AppBar.offsetClassName
 * AppBar.offsetProminentClassName
 * AppBar.offsetDenseClassName
 * AppBar.offsetProminentDenseClassName
 * ```
 *
 * You can also use the provided `rmd-app-bar-offset` mixin to manually apply the offset to one element.
 */
export default class AppBar extends React.Component<IAppBarProps> {
  public static offsetClassName = "rmd-app-bar-offset";
  public static offsetProminentClassName = "rmd-app-bar-offset--prominent";
  public static offsetDenseClassName = "rmd-app-bar-offset--dense";
  public static offsetProminentDenseClassName = "rmd-app-bar-offset--prominent-dense";

  public static propTypes = {
    dense: PropTypes.bool,
    prominent: PropTypes.bool,
    fixed: PropTypes.bool,
    fixedPosition: PropTypes.oneOf(["top", "bottom"]),
    fixedElevation: PropTypes.bool,
    inheritColor: PropTypes.bool,
    theme: PropTypes.oneOf(["primary", "secondary", "default", "clear"]),
  };

  public static defaultProps: IAppBarDefaultProps = {
    fixed: true,
    fixedPosition: "top",
    fixedElevation: true,
    dense: false,
    prominent: false,
    theme: "primary",
  };

  public render() {
    const {
      className,
      dense,
      prominent,
      children,
      fixed,
      fixedPosition,
      fixedElevation,
      theme,
      inheritColor,
      ...props
    } = this.props as AppBarWithDefaultProps;

    const inherit = typeof inheritColor === "boolean" ? inheritColor : theme !== "clear" && theme !== "default";
    return (
      <header
        {...props}
        className={cn(
          `rmd-app-bar rmd-app-bar--${theme}`,
          {
            "rmd-app-bar--child-inherit": inherit,
            "rmd-app-bar--dense": dense && !prominent,
            "rmd-app-bar--prominent": prominent,
            "rmd-app-bar--prominent-dense": dense && prominent,
            "rmd-app-bar--fixed": fixed,
            "rmd-app-bar--fixed-elevation": fixed && fixedElevation,
            [`rmd-app-bar--${fixedPosition}`]: fixed,
          },
          className
        )}
      >
        {children}
      </header>
    );
  }
}
