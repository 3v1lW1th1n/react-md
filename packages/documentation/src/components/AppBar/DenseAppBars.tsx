import * as React from "react";

const DenseAppBars = () => {
  return null;
};

export default DenseAppBars;
