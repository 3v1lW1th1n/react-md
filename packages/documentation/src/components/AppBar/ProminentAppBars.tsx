import * as React from "react";
import { AppBar, AppBarNav, AppBarAction, AppBarTitle, AppBarRow } from "@react-md/app-bar";
import { MenuSVGIcon, FavoriteSVGIcon, SearchSVGIcon, MoreVertSVGIcon } from "@react-md/material-icons";

const ProminentAppBars = () => (
  <React.Fragment>
    <AppBar className="example-group__example" theme="primary" prominent={true} fixed={false}>
      <AppBarRow>
        <AppBarNav>
          <MenuSVGIcon />
        </AppBarNav>
        <AppBarTitle>Page title</AppBarTitle>
        <AppBarAction first={true}>
          <FavoriteSVGIcon />
        </AppBarAction>
        <AppBarAction>
          <SearchSVGIcon />
        </AppBarAction>
        <AppBarAction last={true}>
          <MoreVertSVGIcon />
        </AppBarAction>
      </AppBarRow>
    </AppBar>
    <AppBar className="example-group__example" theme="secondary" prominent={true} fixed={false}>
      <AppBarRow>
        <AppBarTitle keyline={true}>Page title</AppBarTitle>
        <AppBarAction first={true}>
          <FavoriteSVGIcon />
        </AppBarAction>
        <AppBarAction>
          <SearchSVGIcon />
        </AppBarAction>
        <AppBarAction last={true}>
          <MoreVertSVGIcon />
        </AppBarAction>
      </AppBarRow>
    </AppBar>
    <AppBar className="example-group__example" theme="clear" prominent={true} fixed={false}>
      <AppBarRow>
        <AppBarNav>
          <MenuSVGIcon />
        </AppBarNav>
        <AppBarAction first={true}>
          <FavoriteSVGIcon />
        </AppBarAction>
        <AppBarAction>
          <SearchSVGIcon />
        </AppBarAction>
        <AppBarAction last={true}>
          <MoreVertSVGIcon />
        </AppBarAction>
      </AppBarRow>
      <AppBarRow>
        <AppBarTitle keyline={true}>Page title</AppBarTitle>
      </AppBarRow>
    </AppBar>
    <AppBar className="example-group__example" theme="default" prominent={true} fixed={false}>
      <AppBarRow>
        <AppBarAction first={true}>
          <FavoriteSVGIcon />
        </AppBarAction>
        <AppBarAction>
          <SearchSVGIcon />
        </AppBarAction>
        <AppBarAction last={true}>
          <MoreVertSVGIcon />
        </AppBarAction>
      </AppBarRow>
      <AppBarRow>
        <AppBarTitle keyline={true}>Page title</AppBarTitle>
      </AppBarRow>
    </AppBar>
  </React.Fragment>
);

export default ProminentAppBars;
