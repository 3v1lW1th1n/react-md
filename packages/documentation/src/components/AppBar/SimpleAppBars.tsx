import * as React from "react";
import { MenuSVGIcon } from "@react-md/material-icons";
import { AppBar, NavigationButton } from "@react-md/app-bar";
import { Text } from "@react-md/typography";

const SimpleAppBars = () => (
  <React.Fragment>
    <AppBar className="example-group__example">
      <NavigationButton>
        <MenuSVGIcon />
      </NavigationButton>
      <Text type="headline-6">Page title</Text>
    </AppBar>
  </React.Fragment>
);

export default SimpleAppBars;
