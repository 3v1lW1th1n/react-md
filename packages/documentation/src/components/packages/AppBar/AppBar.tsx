import * as React from "react";
import * as Router from "react-router";
import { Switch, Route, Redirect } from "react-router-dom";

import Loadable from "components/Loadable";
import NotFound from "components/NotFound";

const Examples = Loadable(() => import("./Examples"));
const AppBarPropTypes = Loadable(() => import("./AppBarPropTypes"));
const AppBarSassDoc = Loadable(() => import("./AppBarSassDoc"));
const Changelog = Loadable(() => import("./Changelog"));

const AppBar: React.SFC<Router.RouteComponentProps<void>> = ({ match }) => (
  <Switch>
    <Redirect exact={true} from={match.path} to={`${match.path}/examples`} />
    <Route path={`${match.path}/examples`} component={Examples} />
    <Route path={`${match.path}/proptypes`} component={AppBarPropTypes} />
    <Route path={`${match.path}/sassdoc`} component={AppBarSassDoc} />
    <Route path={`${match.path}/changelog`} component={Changelog} />
    <Route component={NotFound} />
  </Switch>
);

export default AppBar;
