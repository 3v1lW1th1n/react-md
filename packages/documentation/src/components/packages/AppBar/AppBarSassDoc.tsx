import * as React from "react";

import SassDocPage from "components/SassDocPage";

// const props = {};

const AppBarSassDoc = () => <SassDocPage name="AppBar" />;

export default AppBarSassDoc;
