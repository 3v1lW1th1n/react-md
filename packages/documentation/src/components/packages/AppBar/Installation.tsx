/* tslint:disable max-line-length */
import * as React from "react";

import { MarkdownPage } from "components/Markdown";

const markdown = `# Installing @react-md/app-bar

## Table of Contents
- [Installing](#installing)
- [Including Styles](#including-styles)
  * [Updating Sass to include \`node_modules\`](#updating-sass-to-include-node_modules)
  * [webpack](#webpack)
  * [create-react-app and node-sass-chokidar](#create-react-app-and-node-sass-chokidar)
  * [Scss File Names and Importing](#scss-file-names-and-importing)
- [Including Styles Quick](#including-styles-quick)

## Installing
\`\`\`sh
$ npm install --save @react-md/app-bar \\
    @react-md/button \\
    @react-md/icon \\
    @react-md/theme \\
    @react-md/typography
\`\`\`

or with \`yarn\`

\`\`\`sh
$ yarn add @react-md/app-bar \\
    @react-md/button \\
    @react-md/icon \\
    @react-md/theme \\
    @react-md/typography
\`\`\`

> NOTE: The packages listed after the first line are required for additional styles. They only need to be added once to your project.



## Including Styles
The following sections will be helpers to update your Sass compiler to include the \`node_modules\` in the \`include-paths\` either for an
app using \`create-react-app\` or a custom webpack config as well as adding \`autoprefixer\`. If you have already completed these steps,
you can skip down to [Installing Styles Quick](#including-styles-quick).

### Updating Sass to include \`node_modules\`
If you want to include the SCSS styles for \`@react-md/app-bar\`, you will need to update your Sass compiler to include the \`node_modules\` in the paths as well as add [autoprefixer](https://github.com/postcss/autoprefixer) to handle multiple browser compatibility.

> If you are using [create-react-app](https://github.com/facebook/create-react-app), the autoprefixer is already included.

### webpack
\`\`\`diff
 {
   test: /.scss$/,
   use: [{
     loader: 'style-loader',
     options: { sourceMap: true },
   }, {
     loader: 'css-loader',
     options: { sourceMap: true, importLoaders: 2 },
   }, {
     loader: 'postcss',
     options: { sourceMap: true },
   }, {
     loader: 'sass-loader',
     options: {
       sourceMap: true,
+      includePaths: [
+        './node_modules', // or whatever relative path it is to node_modules
+      ],
     },
   }],
 }
\`\`\`

### create-react-app and node-sass-chokidar
\`\`\`diff
   "scripts": {
+    "build-css": "node-sass-chokidar --include-path ./node_modules src/ -o src/",
+    "watch-css": "npm run build-csss && npm run build-css -- --watch --recursive"
   }
\`\`\`


### Scss File Names and Importing
This package exports the following files:
- _app-bar.scss
- _mixins.scss
- _variables.scss
- styles.scss

which are mostly self describing about what they export.
 The \`styles.scss\` is the only file that will generate styles by default
when imported into your project as well as including all variables, functions, and mixins for this package. The \`app-bar.scss\` is a file that is used to just automatically import all files within
this package for convenience. So for example:
\`\`\`scss
@import '@react-md/app-bar/dist/variables';
@import '@react-md/app-bar/dist/mixins';
\`\`\`

would be the same as
\`\`\`scss
@import '@react-md/app-bar/dist/app-bar';
\`\`\`


This package also requires the styles from the following packages:
- \`@react-md/button\`
- \`@react-md/icon\`
- \`@react-md/theme\`
- \`@react-md/typography\`

so you will need to follow the above steps to include the styles for them as well.

## Including Styles Quick
\`\`\`scss
@import '@react-md/app-bar/dist/app-bar';
@import '@react-md/button/dist/button';
@import '@react-md/icon/dist/icon';
@import '@react-md/theme/dist/theme';
@import '@react-md/typography/dist/typography';

@include react-md-app-bar;
@include react-md-button;
@include react-md-icon;
@include react-md-theme;
@include react-md-typography;
\`\`\`
`;

const Installation = () => <MarkdownPage markdown={markdown} />;

export default Installation;
