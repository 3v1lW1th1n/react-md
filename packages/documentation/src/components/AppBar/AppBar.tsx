import * as React from "react";

import ExamplesPage from "components/ExamplesPage";

import SimpleAppBars from "./SimpleAppBars";

const examples = [{
  title: "Simple",
  children: <SimpleAppBars />,
}];

const AppBar = () => <ExamplesPage title="App Bar" examples={examples} />;

export default AppBar;
