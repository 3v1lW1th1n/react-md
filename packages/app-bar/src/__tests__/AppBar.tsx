import React from "react";
import { render } from "@testing-library/react";

import AppBar from "../AppBar";

describe("AppBar", () => {
  it("should render as a header element by default", () => {
    const { container } = render(<AppBar />);
    expect(container.querySelector("header")).not.toBeNull();
  });

  it("should be able to render as another component", () => {
    const { container, rerender } = render(<AppBar component="div" />);
    const find = () => container.querySelector(".rmd-app-bar") as HTMLElement;
    expect(find().tagName).toBe("DIV");

    rerender(<AppBar component="header" />);
    expect(find().tagName).toBe("HEADER");

    rerender(<AppBar component="section" />);
    expect(find().tagName).toBe("SECTION");
  });

  it("should apply the primary theme by default", () => {
    const { container } = render(<AppBar />);
    const appBar = container.querySelector("header") as HTMLDivElement;
    expect(appBar.classList.contains("rmd-app-bar--primary")).toBe(true);
  });

  it("should apply the correct theme class name", () => {
    const { container, rerender } = render(<AppBar />);
    const getAppBar = () => container.querySelector("header") as HTMLDivElement;
    const getClassList = () => getAppBar().classList;
    let classList = getClassList();
    expect(classList.contains("rmd-app-bar--primary")).toBe(true);

    rerender(<AppBar theme="primary" />);
    classList = getClassList();
    expect(classList.contains("rmd-app-bar--primary")).toBe(true);
    expect(classList.contains("rmd-app-bar--secondary")).toBe(false);
    expect(classList.contains("rmd-app-bar--default")).toBe(false);

    rerender(<AppBar theme="secondary" />);
    classList = getClassList();
    expect(classList.contains("rmd-app-bar--primary")).toBe(false);
    expect(classList.contains("rmd-app-bar--secondary")).toBe(true);
    expect(classList.contains("rmd-app-bar--default")).toBe(false);

    rerender(<AppBar theme="default" />);
    classList = getClassList();
    expect(classList.contains("rmd-app-bar--primary")).toBe(false);
    expect(classList.contains("rmd-app-bar--secondary")).toBe(false);
    expect(classList.contains("rmd-app-bar--default")).toBe(true);

    // clear adds no additional classes
    rerender(<AppBar theme="clear" />);
    classList = getClassList();
    expect(classList.contains("rmd-app-bar--primary")).toBe(false);
    expect(classList.contains("rmd-app-bar--secondary")).toBe(false);
    expect(classList.contains("rmd-app-bar--default")).toBe(false);
    expect(getAppBar().className).toBe("rmd-app-bar");
  });

  it("should apply the correct dense and prominent class names", () => {
    const { container, rerender } = render(<AppBar />);
    const getClassList = () =>
      (container.querySelector("header") as HTMLDivElement).classList;
    let classList = getClassList();
    expect(classList.contains("rmd-app-bar--dense")).toBe(false);
    expect(classList.contains("rmd-app-bar--prominent")).toBe(false);
    expect(classList.contains("rmd-app-bar--prominent-dense")).toBe(false);

    rerender(<AppBar dense />);
    classList = getClassList();
    expect(classList.contains("rmd-app-bar--dense")).toBe(true);
    expect(classList.contains("rmd-app-bar--prominent")).toBe(false);
    expect(classList.contains("rmd-app-bar--prominent-dense")).toBe(false);

    rerender(<AppBar prominent />);
    classList = getClassList();
    expect(classList.contains("rmd-app-bar--dense")).toBe(false);
    expect(classList.contains("rmd-app-bar--prominent")).toBe(true);
    expect(classList.contains("rmd-app-bar--prominent-dense")).toBe(false);

    rerender(<AppBar prominent dense />);
    classList = getClassList();
    expect(classList.contains("rmd-app-bar--dense")).toBe(false);
    expect(classList.contains("rmd-app-bar--prominent")).toBe(true);
    expect(classList.contains("rmd-app-bar--prominent-dense")).toBe(true);
  });

  it("should default to enabling the fixed elevation class and top classes when fixed is enabled", () => {
    const { container } = render(<AppBar fixed />);
    const getClassList = () =>
      (container.querySelector("header") as HTMLDivElement).classList;
    const classList = getClassList();
    expect(classList.contains("rmd-app-bar--fixed")).toBe(true);
    expect(classList.contains("rmd-app-bar--fixed-elevation")).toBe(true);
    expect(classList.contains("rmd-app-bar--top")).toBe(true);
    expect(classList.contains("rmd-app-bar--bottom")).toBe(false);
  });

  it("should apply the correct fixed class names", () => {
    const { container, rerender } = render(<AppBar />);
    const getClassList = () =>
      (container.querySelector("header") as HTMLDivElement).classList;
    let classList = getClassList();
    expect(classList.contains("rmd-app-bar--fixed")).toBe(false);
    expect(classList.contains("rmd-app-bar--fixed-elevation")).toBe(false);
    expect(classList.contains("rmd-app-bar--top")).toBe(false);
    expect(classList.contains("rmd-app-bar--bottom")).toBe(false);

    rerender(<AppBar fixedPosition="bottom" />);
    classList = getClassList();
    expect(classList.contains("rmd-app-bar--fixed")).toBe(false);
    expect(classList.contains("rmd-app-bar--fixed-elevation")).toBe(false);
    expect(classList.contains("rmd-app-bar--top")).toBe(false);
    expect(classList.contains("rmd-app-bar--bottom")).toBe(false);

    rerender(<AppBar fixed fixedPosition="bottom" />);
    classList = getClassList();
    expect(classList.contains("rmd-app-bar--fixed")).toBe(true);
    expect(classList.contains("rmd-app-bar--fixed-elevation")).toBe(true);
    expect(classList.contains("rmd-app-bar--top")).toBe(false);
    expect(classList.contains("rmd-app-bar--bottom")).toBe(true);

    rerender(<AppBar fixed fixedPosition="top" fixedElevation={false} />);
    classList = getClassList();
    expect(classList.contains("rmd-app-bar--fixed")).toBe(true);
    expect(classList.contains("rmd-app-bar--fixed-elevation")).toBe(false);
    expect(classList.contains("rmd-app-bar--top")).toBe(true);
    expect(classList.contains("rmd-app-bar--bottom")).toBe(false);

    rerender(<AppBar fixed fixedPosition="bottom" fixedElevation={false} />);
    classList = getClassList();
    expect(classList.contains("rmd-app-bar--fixed")).toBe(true);
    expect(classList.contains("rmd-app-bar--fixed-elevation")).toBe(false);
    expect(classList.contains("rmd-app-bar--top")).toBe(false);
    expect(classList.contains("rmd-app-bar--bottom")).toBe(true);
  });

  it("should render children correctly", () => {
    const { container } = render(
      <AppBar>
        <h3>Hello</h3>
      </AppBar>
    );
    expect(container.querySelector("h3")).not.toBeNull();
  });
});
