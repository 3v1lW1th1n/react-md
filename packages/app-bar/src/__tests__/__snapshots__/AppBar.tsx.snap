// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppBar should render child app bars as divs to help with stacked layouts 1`] = `
<div>
  <header
    class="rmd-app-bar rmd-app-bar--primary rmd-app-bar--prominent"
    id="app-bar"
  >
    <div
      class="rmd-app-bar"
      id="app-bar-row-1"
    >
      Here is some content on a row
    </div>
    <div
      class="rmd-app-bar"
      id="app-bar-row-2"
    >
      Second row
    </div>
  </header>
</div>
`;

exports[`AppBar should render correctly (with snapshots) 1`] = `
<div>
  <header
    class="rmd-app-bar rmd-app-bar--primary"
    id="app-bar-id"
  >
    <span>
      Here is some content
    </span>
  </header>
</div>
`;

exports[`AppBar should render correctly (with snapshots) 2`] = `
<div>
  <header
    class="rmd-app-bar rmd-app-bar--secondary rmd-app-bar--fixed rmd-app-bar--top rmd-app-bar--fixed-elevation"
    id="app-bar-id"
  >
    <span>
      Here is some content
    </span>
  </header>
</div>
`;

exports[`AppBar should render correctly (with snapshots) 3`] = `
<div>
  <header
    class="rmd-app-bar rmd-app-bar--secondary rmd-app-bar--fixed rmd-app-bar--top"
    id="app-bar-id"
  >
    <span>
      Here is some content
    </span>
  </header>
</div>
`;
